swift-playground.cpp

```diff
@@ -0,0 +1,288 @@
+#include <iostream>
+#include <string>
+#include <mutex>
+
+#include "CommandLineOptions.h"
+#include "REPL.h"
+#include "Strings.h"
+
+#include <windows.h>
+#include <windowsx.h>
+#include <winuser.h>
+#include <richedit.h>
+#include <namedpipeapi.h>
+#include <io.h>
+#include <fcntl.h>
+
+CommandLineOptions g_opts;
+std::unique_ptr<REPL> g_repl;
+HWND g_window;
+HWND g_recompile_btn;
+HWND g_continue_btn;
+HWND g_text;
+HWND g_output;
+std::string g_prev_text;
+std::string g_curr_text;
+int g_min_line;
+int g_num_lines;
+
+HANDLE g_stdout_write_pipe;
+HANDLE g_stdout_read_pipe;
+std::vector<char> g_output_text;
+std::mutex g_output_text_lock;
```

Comment:
So the thing that concerns me here is that a lot of these globals are used in multiple functions and it's kind of difficult to tell if they ever are used in a concurrent context, i.e. two different threads are trying to use or modify them. You protect `g_output_text` but what about `g_prev_text` or `g_curr_text`? Is it possible that one thread is using `g_curr_text` while another is modifying it? Maybe I'm not familiar enough with how windows desktop applications work but I'm having a hard time figuring out what exactly is safe here.

Reply:
g_output is the only variable shared between threads, so I only protect it. There are two threads: the message pump thread and the thread that reads stdout and prints to the output box.

Reply:
Oh crap also g_output_text